yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -590,6 +605,18 @@ private[spark] class ApplicationMaster(
           case None => logWarning("Container allocator is not ready to kill executors yet.")
         }
         context.reply(true)
+
+      case GetExecutorLossReason(eid) =>
+        Option(allocator) match {
+          case Some(a) =>
+            pendingLossReasonRequests.synchronized {
+              pendingLossReasonRequests
```

Comment:
When would multiple requests for the same executor arrive?

Because if that case is possible, there's a possible race that would lead to a slow leak. If the allocator notifies of the executor reason, you'll reply to the existing requests; but if a new request for that executor's loss reason arrives after that, it will never be replied to, and forever live in `pendingLossReasonRequests`.